### ISSUE TITLE
domish.Element can't easy access elements named 'name'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2429]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jack reported|
|-|-|
|Trac ID|trac#2429|
|Type|enhancement|
|Created|2007-02-01 22:16:09Z|
|Branch|https://github.com/twisted/twisted/tree/domish-qualified-elements-2429-2|

domish.Element.name returns the name of the Element, but this hides any child elements who's name is 'name'.  For a real life example, this means you can't easily access the name child of an iq:version reply.  

The following snippet illustrates the problem:
```
from twisted.words.xish import domish

elem = domish.Element((None, 'query'))
elem.addElement('name', content='asdf')

if str(elem.name) == 'name':
  print "works"
else:
  print "doesn't work"
```

The workaround appears to be to scan through Element.elements.  Note that this wouldn't be so bad if there was an Element.getChildByName or something.

<details><summary>Searchable metadata</summary>

```
trac-id__2429 2429
type__enhancement enhancement
reporter__jack jack
priority__normal normal
milestone__Twisted_8_2 Twisted-8.2
branch__branches_domish_qualified_elements_2429_2 branches/domish-qualified-elements-2429-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1170368169000000 1170368169000000
changetime__1225311453000000 1225311453000000
version__None None
owner__ 
cc__jack cc__ralphm cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm set owner to @ralphm|
|-|-|

Indeed, walking Element.elements() is currently the only way to do this. Faking attributes for less verbose code has its drawbacks. But, I think it is reasonable to define a new API for that.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

Created a new branch [domish-qualified-elements-2429](domish-qualified-elements-2429) that adds two arguments to an `Element`'s `elements` method: `uri` and `name`. If specified, it will only return elements by that qualified name. How does that sound?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to jack|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @ralphm|
|-|-|

It would be nice if the test actually used the 'name' attribute as the original issue here described.

Longer-term, we should probably plan to use one of the fancy, fast, new XML APIs, e.g. lxml, no?  Is there a plan for the future of domish?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm set owner to @glyph|
|-|-|

Replying to [glyph](#note_5):
> It would be nice if the test actually used the 'name' attribute as the original issue here described.

Sure, done. Does that conclude your review?

> Longer-term, we should probably plan to use one of the fancy, fast, new XML APIs, e.g. lxml, no?  Is there a plan for the future of domish?

This has been raised so many times now that I should make a FAQ out of it. In short: the processing of streaming XML as defined in the XMPP context brings a number of requirements for serialization and DOM usage that are simply not met by any of the existing XML libraries out there, regardless of their fancyness.

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|jack commented|
|-|-|

> > Longer-term, we should probably plan to use one of the fancy, fast, new XML APIs, e.g. lxml, no?  Is there a plan for the future of domish?
>
> This has been raised so many times now that I should make a FAQ out of it. In short: the processing of streaming XML as defined in the XMPP context brings a number of requirements for serialization and DOM usage that are simply not met by any of the existing XML libraries out there, regardless of their fancyness.

Note that we made an effort to move to lxml and found that the added overhead of going through python made it about the same speed as the pure python version.  Perhaps we made a mistake along the way, but this effort took several weeks and produced no results for us.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [jack](#note_7):

> Note that we made an effort to move to lxml and found that the added overhead of going through python made it about the same speed as the pure python version.  Perhaps we made a mistake along the way, but this effort took several weeks and produced no results for us.

The reason I raised this was not so much the speed issue as the maintenance / integration issue.  Assuming we can have the same functionality, it would be better to rely on a well-used third party library than to stick with something custom.  There's likely to be more third-party code that will work on an lxml data-structure than a xish one.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [ralphm](#note_6):
> Replying to [glyph](#note_5):
> > It would be nice if the test actually used the 'name' attribute as the original issue here described.
>
> Sure, done. Does that conclude your review?

Yes.

It looked like the ticket wasn't really in review, but was waiting for some feedback from Jack to determine whether it actually addressed his issue.  Based on the tests I couldn't tell whether it did.

> > Longer-term, we should probably plan to use one of the fancy, fast, new XML APIs, e.g. lxml, no?  Is there a plan for the future of domish?
>
> This has been raised so many times now that I should make a FAQ out of it. In short: the processing of streaming XML as defined in the XMPP context brings a number of requirements for serialization and DOM usage that are simply not met by any of the existing XML libraries out there, regardless of their fancyness.

A FAQ might be a good idea, but before that, an explanation in xish's docstring of what these mythical requirements actually _are_ would be better :).  I don't have the XMPP experience to understand what makes xish worthwhile.  This snippet from `domish.Element`'s docstring, for example, suggests that it is just like any other XML API...
```
An Element contains a series of attributes (name/value pairs), content
(character data), and other child Element objects. When building a document
with markup (such as HTML or XML), use this object as the starting point.
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

On second thought, including a full discussion of this issue here is probably just going to clutter up this (relatively small) issue.

I've created [#3201](https://github.com/twisted/twisted/issues/3201) to deal with the discussion of xish's future.

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|jack commented|
|-|-|

This patch appears to be fine.  It doesn't really help our particular case much since we are looping through a dict of keys trying to extract elements.  Using this does a double loop, and our "workaround" for the bug is more efficient already.  However this does handle the singleton cases quite nicely.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @ralphm|
|-|-|

  * docstrings for the tests could be improved.  Instead of _Test getting child elements from an Element._, how about something like _L{Element.elements} returns an iterable of that L{Element}'s children._?
  * the `IElement.addElement` method has a very confusing, seemingly overly complex signature.  I mention this since I looked at it to compare the new signature of `Element.elements`.  Why the inconsistency in URI specification?  And why does `IElement.addElement` accept an URI in two different ways (requiring an isinstance in the process, which, by the way, would be more simply written `isinstance(name, tuple)`)?
  * I see `elements` isn't part of `IElement`.  Any particular reason for this?  I'm not sure I know what purpose `IElement` is serving in this code, anyway.  Is the intent for third-parties to be able to supply their own implementations?
  * I think Elements can have non-element children?  If so, there should be a test for the behavior of `elements` in that case.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_13"></a>|@ralphm set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_12):
>   * docstrings for the tests could be improved.  Instead of _Test getting child elements from an Element._, how about something like _L{Element.elements} returns an iterable of that L{Element}'s children._?

I'm not sure if there are guidelines towards docstrings for tests, but I try to describe what a method does (whether a test method or regular code), and your example describes does not seem to fit that pattern. I do agree that it could be more clear.

>   * the `IElement.addElement` method has a very confusing, seemingly overly complex signature.  I mention this since I looked at it to compare the new signature of `Element.elements`.  Why the inconsistency in URI specification?  And why does `IElement.addElement` accept an URI in two different ways (requiring an isinstance in the process, which, by the way, would be more simply written `isinstance(name, tuple)`)?

Apart from historical choices, I want to note at first that the namespace URI of the qualified name of an element is not always the same as the default namespace URI. The `name` parameter is either qualified name, or the local name, whereas `defaultUri` is the default namespace URI, and, if specified, also the namespace uri of the element if `name` is only a local name.

As such, this interface allows for different ways to add an element:

  * `e.addElement('name')`: add a child element with local name `name` in `e`'s default namespace. The default namespace is inherited from the parent `e`.
  * `e.addElement(('myns', 'name'))`: add a child element `{myns}name` (using Clark's notation). The default namespace is inherited from the parent.
  * `e.addElement('name', 'myns')`: identical to the previous.
  * `e.addElement(('myns', 'name'), 'otherns')`: add a child element `{myns}name`, but with `otherns` as the default namespace.

I never use the latter two forms, but believe it must continue to be possible to do the equivalent of the last one.

You are right about the type check, but this part of the code hasn't been touched in a while, I think.

I was tempted to reverse the arguments of `elements` for consistency with `addElement`'s interface, but that would require that you must pass the namespace uri as a keyword argument if you want to have all elements in that namespace returned (not supplying the `name` argument.

>   * I see `elements` isn't part of `IElement`.  Any particular reason for this?  I'm not sure I know what purpose `IElement` is serving in this code, anyway.  Is the intent for third-parties to be able to supply their own implementations?

I didn't think of adding it there. And yes, that was the intent. Among other things, to allow for objects that represent a partial DOM on their own, and/or lazy evaluation.

>   * I think Elements can have non-element children?  If so, there should be a test for the behavior of `elements` in that case.

They can have CDATA children, represented as unicode objects, or `SerializedXML` objects, that contain an already serialized partial DOM in a UTF-8 encoded string.

> Thanks.

Thank you!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @ralphm|
|-|-|

It's redundant to say "Test that ..." at the beginning of each test method's docstring, so I'd prefer that we didn't make that the convention.  If you want to, fine.  However, what comes _after_ that should be more descriptive than what is in the docstrings in this branch.  Where possible, it should include the API which is being tested, and it should include specific details about the API which are of importance.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_16"></a>|@ralphm removed owner|
|-|-|

Not sure why it took me so long, but I adjusted the docstrings.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @ralphm|
|-|-|

Conflict in `test_domish.py`, can you merge forward?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_18"></a>|@ralphm commented|
|-|-|

(In [25179]) Branching to 'domish-qualified-elements-2429-2'

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_19"></a>|@ralphm set owner to @exarkun|
|-|-|

Merged forward.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @ralphm|
|-|-|

  1. `elements` documents its parameters as being of type `unicode`, but the test passes `str`.
  1. Likewise, the attributes of `Element` should be `unicode` but the test passes in `str`.
  1. `C{generator}` doesn't make sense, since _generator_ isn't code.  I'd suggest documenting the return type as an iterator instead.
  1. There should be two blank lines between `test_elements` and `test_elementsWithQN`.

Building at <http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/domish-qualified-elements-2429-2&num_builds=1>.  Once the above points are fixed, if the tests pass, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_22"></a>|@ralphm set status to `closed`|
|-|-|

(In [25184]) Allow for iterating over domish child elements with a specific qualified name.

Author: ralphm.
Reviewer: glyph, exarkun.
Fixes [#2429](https://github.com/twisted/twisted/issues/2429).

This adds two parameters to `twisted.words.xish.domish.Element.elements`, `uri`
and `name`, that restrict the returned child elements to the given qualified
name.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
